File: packages/persistence/src/redaction/receipt.ts
/**
 * The bound redaction receipt of spec/redaction-semantics.md Section 3.3.
 *
 * A receipt is only proof when the guard deterministically replays the transform
 * from the immutable pre-transform snapshot and byte-compares the exact
 * persisted result. A caller-provided boolean or receipt without that comparison
 * is not proof of redaction.
 */

import type { CaptureSinkClass, CaptureSourceClass, GuardFailure } from "./codes.js";
import { hmacSha256Hex } from "./crypto.js";
import { canonicalTagged } from "./durable-json.js";
import type { AuthorityScope } from "./protected-store.js";
import { canonicalJsonString, compareUnicodeCodePoints } from "./portable.js";
import { redactionTransform, type ReplacementMode } from "./pointer.js";
import { ruleSetHash, type CapturePolicy, type RedactionRule } from "./policy.js";
import { SECTION_11_LIMITS, type PortableLimits } from "./limits.js";

export const REDACTION_RECEIPT_API_VERSION =
  "graphengineering.reacher-z.github.io/redaction-receipt/v1alpha2" as const;

/**
 * Section 1.2 row 1 and Sections 2.2/2.3: these eight authoritative classes are
 * "protect, never redact", so no receipt may name one.
 */
export const NEVER_REDACTABLE_SOURCE_CLASSES: readonly string[] = Object.freeze([
  "bound-node-input",
  "checkpoint-state",
  "event-data",
  "graph-input",
  "graph-output",
  "node-output",
  "node-result",
  "run-result",
]);

/** Section 2.3: encryption is never redaction; the protected-store family is excluded. */
export const NEVER_REDACTABLE_SINKS: readonly string[] = Object.freeze([
  "protected-blob-final",
  "protected-blob-memory",
  "protected-blob-temporary",
]);

export interface RedactionReceipt {
  readonly apiVersion: typeof REDACTION_RECEIPT_API_VERSION;
  readonly policyHash: string;
  readonly sourceHash: string;
  readonly resultHash: string;
  readonly ruleSetHash: string;
  readonly transform: "json-pointer-rules/v1alpha2";
  readonly transformImplementationHash: string;
  readonly ruleRegistryHash: string;
  readonly ruleRegistryVersion: number;
  readonly ruleResolutionId: string;
  readonly authorityBindingHash: string;
  readonly tenantScopeHash: string;
  readonly sourceClass: CaptureSourceClass;
  readonly sink: CaptureSinkClass;
  readonly decisionId: string;
  readonly runId: string;
  readonly graphRevision: number;
  readonly occurrenceKind: "event" | "checkpoint" | "sink-write";
  readonly occurrenceId: string;
  readonly occurrenceSequence: number;
  readonly occurredAt: string;
  readonly fieldPath: string;
  readonly paths: readonly string[];
  readonly replacementMode: ReplacementMode;
  readonly count: number;
}

/** Section 3.3 keyed source digest, domain `redaction-source/v1alpha2`. */
export function computeSourceHash(
  identityKey: Uint8Array,
  input: {
    readonly sourceClass: string;
    readonly sink: string;
    readonly fieldPath: string;
    readonly sourceSnapshot: unknown;
  },
): string {
  return hmacSha256Hex(
    identityKey,
    canonicalTagged([
      "redaction-source/v1alpha2",
      input.sourceClass,
      input.sink,
      input.fieldPath,
      input.sourceSnapshot,
    ]),
  );
}

/** Section 3.3 keyed result digest, domain `redaction-result/v1alpha2`. */
export function computeResultHash(
  identityKey: Uint8Array,
  input: {
    readonly sourceClass: string;
    readonly sink: string;
    readonly fieldPath: string;
    readonly result: unknown;
  },
): string {
  return hmacSha256Hex(
    identityKey,
    canonicalTagged([
      "redaction-result/v1alpha2",
      input.sourceClass,
      input.sink,
      input.fieldPath,
      input.result,
    ]),
  );
}

export interface TransformOccurrence {
  readonly decisionId: string;
  readonly runId: string;
  readonly graphRevision: number;
  readonly occurrenceKind: "event" | "checkpoint" | "sink-write";
  readonly occurrenceId: string;
  readonly occurrenceSequence: number;
  readonly occurredAt: string;
  readonly fieldPath: string;
  readonly ruleResolutionId: string;
}

export interface ReceiptRequest {
  readonly identityKey: Uint8Array;
  readonly policy: CapturePolicy;
  readonly policyHash: string;
  readonly rule: RedactionRule;
  readonly sourceClass: CaptureSourceClass;
  readonly sink: CaptureSinkClass;
  readonly occurrence: TransformOccurrence;
  readonly scope: AuthorityScope;
  readonly authorityBindingHash: string;
  readonly tenantScopeHash: string;
  /** The immutable pre-transform snapshot. Never an already-transformed value. */
  readonly sourceSnapshot: unknown;
  readonly limits?: PortableLimits;
}

export type ReceiptResult =
  | {
      readonly valid: true;
      readonly receipt: RedactionReceipt;
      readonly result: unknown;
      readonly canonicalResult: string;
    }
  | { readonly valid: false; readonly failure: GuardFailure };

function denial(code: GuardFailure["code"], reason: string): ReceiptResult {
  return { valid: false, failure: { code, phase: "receipt", reason } };
}

/**
 * Apply the rule to the immutable snapshot, build the receipt, then replay the
 * transform from the same snapshot and byte-compare the canonical result before
 * the write can be authorized (Section 3.3.1 step 6).
 */
export function buildRedactionReceipt(request: ReceiptRequest): ReceiptResult {
  const limits = request.limits ?? SECTION_11_LIMITS;

  if (NEVER_REDACTABLE_SOURCE_CLASSES.includes(request.sourceClass)) {
    return denial("REDACTION_RECEIPT_INVALID", "authoritative-source-class-cannot-be-redacted");
  }
  if (NEVER_REDACTABLE_SINKS.includes(request.sink)) {
    return denial("REDACTION_RECEIPT_INVALID", "protected-store-sink-cannot-carry-a-receipt");
  }
  if (request.rule.sink !== request.sink) {
    return denial("REDACTION_RECEIPT_INVALID", "rule-sink-does-not-match-write-sink");
  }
  if (request.rule.registryVersion !== request.policy.ruleRegistryVersion) {
    return {
      valid: false,
      failure: {
        code: "CAPTURE_POLICY_MISMATCH",
        phase: "receipt",
        reason: "rule-registry-version-changed-during-transform",
      },
    };
  }

  const first = redactionTransform(
    request.sourceSnapshot,
    request.rule.paths,
    request.rule.replacementMode,
    limits,
  );
  if (!first.valid) {
    return { valid: false, failure: { code: first.code, phase: "receipt", reason: first.reason } };
  }

  const paths = [...request.rule.paths];
  for (let index = 1; index < paths.length; index += 1) {
    if (compareUnicodeCodePoints(paths[index - 1] as string, paths[index] as string) >= 0) {
      return denial("REDACTION_RECEIPT_INVALID", "receipt-paths-are-not-in-strict-order");
    }
  }

  const receipt: RedactionReceipt = {
    apiVersion: REDACTION_RECEIPT_API_VERSION,
    policyHash: request.policyHash,
    sourceHash: computeSourceHash(request.identityKey, {
      sourceClass: request.sourceClass,
      sink: request.sink,
      fieldPath: request.occurrence.fieldPath,
      sourceSnapshot: request.sourceSnapshot,
    }),
    resultHash: computeResultHash(request.identityKey, {
      sourceClass: request.sourceClass,
      sink: request.sink,
      fieldPath: request.occurrence.fieldPath,
      result: first.output,
    }),
    ruleSetHash: ruleSetHash([request.rule]),
    transform: "json-pointer-rules/v1alpha2",
    transformImplementationHash: request.policy.transformImplementationHash,
    ruleRegistryHash: request.policy.ruleRegistryHash,
    ruleRegistryVersion: request.policy.ruleRegistryVersion,
    ruleResolutionId: request.occurrence.ruleResolutionId,
    authorityBindingHash: request.authorityBindingHash,
    tenantScopeHash: request.tenantScopeHash,
    sourceClass: request.sourceClass,
    sink: request.sink,
    decisionId: request.occurrence.decisionId,
    runId: request.occurrence.runId,
    graphRevision: request.occurrence.graphRevision,
    occurrenceKind: request.occurrence.occurrenceKind,
    occurrenceId: request.occurrence.occurrenceId,
    occurrenceSequence: request.occurrence.occurrenceSequence,
    occurredAt: request.occurrence.occurredAt,
    fieldPath: request.occurrence.fieldPath,
    paths,
    replacementMode: request.rule.replacementMode,
    count: paths.length,
  };
  if (receipt.count !== receipt.paths.length) {
    return denial("REDACTION_RECEIPT_INVALID", "count-does-not-equal-paths-length");
  }

  // Deterministic replay from the same immutable snapshot, byte-compared.
  const replay = redactionTransform(
    request.sourceSnapshot,
    receipt.paths,
    receipt.replacementMode,
    limits,
  );
  if (!replay.valid) {
    return denial("REDACTION_RECEIPT_INVALID", "replay-denied-after-first-application");
  }
  const canonicalResult = canonicalJsonString(first.output);
  if (canonicalJsonString(replay.output) !== canonicalResult) {
    return denial("REDACTION_RECEIPT_INVALID", "replay-does-not-reproduce-the-candidate");
  }

  return { valid: true, receipt, result: first.output, canonicalResult };
}

/**
 * Verify a claimed receipt against the immutable snapshot and a persisted
 * result. Changing source, result, rule order, registry, transform identity,
 * authority, tenant, sink, occurrence, timestamp, or field path while retaining
 * an old receipt yields `REDACTION_RECEIPT_INVALID`.
 */
export function verifyRedactionReceipt(
  request: ReceiptRequest & { readonly claimed: RedactionReceipt; readonly persistedResult: unknown },
): { readonly valid: true } | { readonly valid: false; readonly failure: GuardFailure } {
  const rebuilt = buildRedactionReceipt(request);
  if (!rebuilt.valid) return rebuilt;
  if (canonicalTagged(rebuilt.receipt) !== canonicalTagged(request.claimed)) {
    return {
      valid: false,
      failure: {
        code: "REDACTION_RECEIPT_INVALID",
        phase: "receipt",
        reason: "receipt-does-not-bind-this-occurrence",
      },
    };
  }
  if (canonicalJsonString(request.persistedResult) !== rebuilt.canonicalResult) {
    return {
      valid: false,
      failure: {
        code: "REDACTION_RECEIPT_INVALID",
        phase: "receipt",
        reason: "persisted-result-differs-from-replay",
      },
    };
  }
  return { valid: true };
}

File: python/src/graph_engineering/redaction/receipt.py
"""The bound redaction receipt of Section 3.3.

A receipt is proof of one exact transform occurrence.  It is constructed only
from a completed transform, and it is accepted only when the guard
deterministically replays that transform from the immutable pre-transform
snapshot and byte-compares the canonical result.  A caller-provided boolean or a
receipt without that comparison is not proof of redaction.
"""

from __future__ import annotations

import hashlib
import hmac
import re
from collections.abc import Mapping, Sequence
from dataclasses import dataclass
from typing import Final, Literal, TypeAlias

from ..canonical import canonical_bytes
from ..models import JsonValue
from .errors import RedactionFailure, failure
from .inventory import (
    NEVER_REDACTABLE_SINKS,
    NEVER_REDACTABLE_SOURCE_CLASSES,
    sink_row,
    source_row,
)
from .limits import DEFAULT_LIMITS, MAX_POINTERS_PER_RULE, PortableLimits
from .pointer import (
    PointerSyntaxError,
    PointerTransformResult,
    apply_pointer_transform,
    decode_pointer,
)
from .policy import RedactionRule
from .protect import canonical_tagged

REDACTION_RECEIPT_API_VERSION: Final = (
    "graphengineering.reacher-z.github.io/redaction-receipt/v1alpha2"
)
REDACTION_TRANSFORM: Final = "json-pointer-rules/v1alpha2"
SOURCE_DOMAIN: Final = "redaction-source/v1alpha2"
RESULT_DOMAIN: Final = "redaction-result/v1alpha2"

_SHA256 = re.compile(r"^[0-9a-f]{64}$")
_IDENTIFIER = re.compile(r"^[A-Za-z0-9][A-Za-z0-9._-]{0,127}$")
_TIMESTAMP = re.compile(
    r"^\d{4}-(0[1-9]|1[0-2])-(0[1-9]|[12]\d|3[01])"
    r"T([01]\d|2[0-3]):[0-5]\d:[0-5]\d(?:\.\d+)?(?:Z|[+-](?:[01]\d|2[0-3]):[0-5]\d)$"
)

OccurrenceKind: TypeAlias = Literal["event", "checkpoint", "sink-write"]
_OCCURRENCE_KINDS: Final = ("event", "checkpoint", "sink-write")

RECEIPT_REQUIRED_FIELDS: Final[tuple[str, ...]] = (
    "apiVersion",
    "policyHash",
    "sourceHash",
    "resultHash",
    "ruleSetHash",
    "transform",
    "transformImplementationHash",
    "ruleRegistryHash",
    "ruleRegistryVersion",
    "ruleResolutionId",
    "authorityBindingHash",
    "tenantScopeHash",
    "sourceClass",
    "sink",
    "decisionId",
    "runId",
    "graphRevision",
    "occurrenceKind",
    "occurrenceId",
    "occurrenceSequence",
    "occurredAt",
    "fieldPath",
    "paths",
    "replacementMode",
    "count",
)


@dataclass(frozen=True, slots=True)
class TransformOccurrence:
    """The write context a receipt must equal exactly."""

    source_class: str
    sink: str
    decision_id: str
    run_id: str
    graph_revision: int
    occurrence_kind: OccurrenceKind
    occurrence_id: str
    occurrence_sequence: int
    occurred_at: str
    field_path: str


@dataclass(frozen=True, slots=True)
class RedactionReceipt:
    """The closed proof object."""

    policy_hash: str
    source_hash: str
    result_hash: str
    rule_set_hash: str
    transform_implementation_hash: str
    rule_registry_hash: str
    rule_registry_version: int
    rule_resolution_id: str
    authority_binding_hash: str
    tenant_scope_hash: str
    occurrence: TransformOccurrence
    paths: tuple[str, ...]
    replacement_mode: Literal["remove", "constant-token"]

    @property
    def count(self) -> int:
        return len(self.paths)

    def as_document(self) -> dict[str, JsonValue]:
        return {
            "apiVersion": REDACTION_RECEIPT_API_VERSION,
            "policyHash": self.policy_hash,
            "sourceHash": self.source_hash,
            "resultHash": self.result_hash,
            "ruleSetHash": self.rule_set_hash,
            "transform": REDACTION_TRANSFORM,
            "transformImplementationHash": self.transform_implementation_hash,
            "ruleRegistryHash": self.rule_registry_hash,
            "ruleRegistryVersion": self.rule_registry_version,
            "ruleResolutionId": self.rule_resolution_id,
            "authorityBindingHash": self.authority_binding_hash,
            "tenantScopeHash": self.tenant_scope_hash,
            "sourceClass": self.occurrence.source_class,
            "sink": self.occurrence.sink,
            "decisionId": self.occurrence.decision_id,
            "runId": self.occurrence.run_id,
            "graphRevision": self.occurrence.graph_revision,
            "occurrenceKind": self.occurrence.occurrence_kind,
            "occurrenceId": self.occurrence.occurrence_id,
            "occurrenceSequence": self.occurrence.occurrence_sequence,
            "occurredAt": self.occurrence.occurred_at,
            "fieldPath": self.occurrence.field_path,
            "paths": list(self.paths),
            "replacementMode": self.replacement_mode,
            "count": self.count,
        }


def rule_set_hash(rules: Sequence[RedactionRule]) -> str:
    """SHA-256 of canonical Tagged Durable JSON for the ordered resolved rules."""

    ordered: list[JsonValue] = [
        {
            "ruleId": rule.rule_id,
            "registryVersion": rule.registry_version,
            "sink": rule.sink,
            "paths": list(rule.paths),
            "replacementMode": rule.replacement_mode,
        }
        for rule in rules
    ]
    return hashlib.sha256(canonical_tagged(ordered)).hexdigest()


def source_hash(
    identity_key: bytes,
    *,
    source_class: str,
    sink: str,
    field_path: str,
    source_snapshot: JsonValue,
) -> str:
    return hmac.new(
        identity_key,
        canonical_tagged([SOURCE_DOMAIN, source_class, sink, field_path, source_snapshot]),
        hashlib.sha256,
    ).hexdigest()


def result_hash(
    identity_key: bytes,
    *,
    source_class: str,
    sink: str,
    field_path: str,
    transformed: JsonValue,
) -> str:
    return hmac.new(
        identity_key,
        canonical_tagged([RESULT_DOMAIN, source_class, sink, field_path, transformed]),
        hashlib.sha256,
    ).hexdigest()


def validate_receipt_document(
    document: object,
    *,
    limits: PortableLimits = DEFAULT_LIMITS,
) -> RedactionFailure | None:
    """Closed structural plus semantic receipt validation."""

    invalid = _receipt_invalid
    if not isinstance(document, Mapping):
        return invalid()
    if set(document.keys()) != set(RECEIPT_REQUIRED_FIELDS):
        return invalid()
    if document["apiVersion"] != REDACTION_RECEIPT_API_VERSION:
        return invalid()
    if document["transform"] != REDACTION_TRANSFORM:
        return invalid()
    for field_name in (
        "policyHash",
        "sourceHash",
        "resultHash",
        "ruleSetHash",
        "transformImplementationHash",
        "ruleRegistryHash",
        "authorityBindingHash",
        "tenantScopeHash",
    ):
        value = document[field_name]
        if type(value) is not str or _SHA256.fullmatch(value) is None:
            return invalid()
    for field_name in ("ruleResolutionId", "decisionId", "runId", "occurrenceId"):
        value = document[field_name]
        if type(value) is not str or _IDENTIFIER.fullmatch(value) is None:
            return invalid()
    for field_name, minimum in (("ruleRegistryVersion", 1), ("graphRevision", 1)):
        value = document[field_name]
        if type(value) is not int or value < minimum:
            return invalid()
    sequence = document["occurrenceSequence"]
    if type(sequence) is not int or sequence < 0:
        return invalid()
    occurred_at = document["occurredAt"]
    if type(occurred_at) is not str or _TIMESTAMP.fullmatch(occurred_at) is None:
        return invalid()
    if document["occurrenceKind"] not in _OCCURRENCE_KINDS:
        return invalid()
    if document["replacementMode"] not in ("remove", "constant-token"):
        return invalid()

    source_class = document["sourceClass"]
    if type(source_class) is not str or source_row(source_class) is None:
        return invalid()
    if source_class in NEVER_REDACTABLE_SOURCE_CLASSES:
        # Section 2.2: redacted data can never be authoritative.
        return invalid()
    sink = document["sink"]
    if type(sink) is not str or sink_row(sink) is None:
        return invalid()
    if sink in NEVER_REDACTABLE_SINKS:
        # Section 2.3: encryption is not redaction.
        return invalid()

    field_path = document["fieldPath"]
    try:
        decode_pointer(field_path, limits=limits)
    except PointerSyntaxError:
        return invalid()

    paths = document["paths"]
    if type(paths) is not list or not paths or len(paths) > MAX_POINTERS_PER_RULE:
        return invalid()
    for index, path in enumerate(paths):
        try:
            decode_pointer(path, limits=limits)
        except PointerSyntaxError:
            return invalid()
        if index and not paths[index - 1] < path:
            # Strictly increasing Unicode code-point order after normalization.
            return invalid()
    if document["count"] != len(paths):
        return invalid()
    return None


def build_receipt(
    *,
    identity_key: bytes,
    policy_hash: str,
    transform_implementation_hash: str,
    rule_registry_hash: str,
    rule_registry_version: int,
    rule_resolution_id: str,
    authority_binding_hash: str,
    tenant_scope_hash: str,
    occurrence: TransformOccurrence,
    rules: Sequence[RedactionRule],
    source_snapshot: JsonValue,
    transformed: JsonValue,
    paths: Sequence[str],
    replacement_mode: Literal["remove", "constant-token"],
) -> RedactionReceipt:
    """Construct the receipt from a completed transform."""

    return RedactionReceipt(
        policy_hash=policy_hash,
        source_hash=source_hash(
            identity_key,
            source_class=occurrence.source_class,
            sink=occurrence.sink,
            field_path=occurrence.field_path,
            source_snapshot=source_snapshot,
        ),
        result_hash=result_hash(
            identity_key,
            source_class=occurrence.source_class,
            sink=occurrence.sink,
            field_path=occurrence.field_path,
            transformed=transformed,
        ),
        rule_set_hash=rule_set_hash(rules),
        transform_implementation_hash=transform_implementation_hash,
        rule_registry_hash=rule_registry_hash,
        rule_registry_version=rule_registry_version,
        rule_resolution_id=rule_resolution_id,
        authority_binding_hash=authority_binding_hash,
        tenant_scope_hash=tenant_scope_hash,
        occurrence=occurrence,
        paths=tuple(paths),
        replacement_mode=replacement_mode,
    )


def verify_receipt(
    receipt: RedactionReceipt,
    *,
    identity_key: bytes,
    source_snapshot: JsonValue,
    persisted_result: JsonValue,
    occurrence: TransformOccurrence,
    limits: PortableLimits = DEFAULT_LIMITS,
) -> RedactionFailure | None:
    """Deterministically replay the transform and byte-compare the result.

    Changing source, result, rule order, registry, transform identity,
    authority, tenant, sink, occurrence, timestamp, or field path while retaining
    an old receipt yields ``REDACTION_RECEIPT_INVALID``.
    """

    structural = validate_receipt_document(receipt.as_document(), limits=limits)
    if structural is not None:
        return structural
    if receipt.occurrence != occurrence:
        return _receipt_invalid()

    replayed = apply_pointer_transform(
        source_snapshot,
        list(receipt.paths),
        receipt.replacement_mode,
        limits=limits,
    )
    if not isinstance(replayed, PointerTransformResult):
        return replayed.failure
    if canonical_bytes(replayed.output) != canonical_bytes(persisted_result):
        return _receipt_invalid()

    expected_source = source_hash(
        identity_key,
        source_class=occurrence.source_class,
        sink=occurrence.sink,
        field_path=occurrence.field_path,
        source_snapshot=source_snapshot,
    )
    expected_result = result_hash(
        identity_key,
        source_class=occurrence.source_class,
        sink=occurrence.sink,
        field_path=occurrence.field_path,
        transformed=replayed.output,
    )
    if not hmac.compare_digest(expected_source, receipt.source_hash):
        return _receipt_invalid()
    if not hmac.compare_digest(expected_result, receipt.result_hash):
        return _receipt_invalid()
    return None


def _receipt_invalid() -> RedactionFailure:
    return failure("REDACTION_RECEIPT_INVALID", "receipt")
